Enable "strict" mode for runtime code.

maxmellen opened the pull request:
Addresses #3324

This PR picks up where #3426 left off setting TypeScript's strict mode on by default for runtime code and updates the standard library and tests to support those changes.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/denoland/deno?pullRequest=3899) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/denoland/deno?pullRequest=3899) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: kitsonk<br/>:x: maxmellen<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/denoland/deno?pullRequest=3899) it.</sub>

maxmellen commented:
@kitsonk I finally got the CI checks to pass.

Do you think you could have a look at the changes when you get the chance?
I don't mind keeping this branch up to date with upstream changes but I'd like to get an idea if this has any chance of being merged before putting anymore time into it.

kitsonk commented on cli/js/buffer_test.ts:
```diff
@@ -78,8 +78,8 @@ function repeat(c: string, bytes: number): Uint8Array {
   init();
-  const buf = new Buffer(testBytes.buffer as ArrayBuffer);
-  check(buf, testString);
+  const buf = new Buffer(testBytes!.buffer as ArrayBuffer);
```
We should use `assert(testBytes)` here so the not null assertion isn't required.

maxmellen replied:
I only now learned about TypeScript 3.7's "assertion functions" 🙂
I'll update my PR to take advantage of that wherever possible.

kitsonk commented on cli/js/chmod_test.ts:
```diff
@@ -16,7 +16,7 @@ testPerm({ read: true, write: true }, function chmodSyncSuccess(): void {
   if (isNotWindows) {
     const fileInfo = Deno.statSync(filename);
-    assertEquals(fileInfo.mode & 0o777, 0o777);
+    assertEquals(fileInfo.mode! & 0o777, 0o777);
```
And again, we should use `assert(fileInfo.mode)`

kitsonk commented on cli/js/console_test.ts:
```diff
@@ -14,8 +14,8 @@ const customInspect = Deno.symbols.customInspect;
   stringifyArgs
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-} = Deno[Deno.symbols.internal] as any;
+  // @ts-ignore
```
Was there another reason to change this?  We should keep the `any`.

maxmellen replied:
The issue here is that with the `noImplicitAny` setting set to true (implicitly turned on by the "strict" flag), TypeScript won't allow indexing `Deno` with `Deno.symbols.internal` without adding that symbol as an index in `Deno`'s type declaration.

Another issue there is that I'm not sure that it's possible to add a computed property to a namespace. The only way I could think of to make this work would be to change `Deno` into a `const` like in [this example](http://www.typescriptlang.org/play/index.html?ssl=1&ssc=1&pln=13&pc=37#code/PTAEEEBNIZ1BLAdpeAnApgYwC7wPaKjZ6gCGAbnvJKJmpgK4A2pqoGAZuhopugoQBiePAChIWFhlCJSAW3QwADqT6gA+jACecgEZ4mcAN6jQZ0OgAeSvKmy0CMe7tYAuUA0TwAjg37a9AwBuUQBfUXFJVn5MR3thPHcTc3Z0UkgCJi1QAP1Dd2wtJXQ8Dg1cgxgQlIBtTR08mAA6F1QAXXdEBj1uEPDIzCl+Dk8cfEJVPiVsGAA5bt1uAApEToXuAEp3SmoQ0Un0abn11CWEmoSmisMW1jaNoKA). (Note this example will only type check but won't run.)

Do you think it's worth updating the `Deno` type declaration as described? I'm a bit conflicted because I think it makes sense for `Deno` to be a namespace and I suppose this internal symbol wouldn't be used by user code. What do you think?

kitsonk commented on cli/js/console_test.ts:
```diff
@@ -306,6 +306,7 @@ test(function consoleTestCallToStringOnLabel(): void {
   for (const method of methods) {
     let hasCalled = false;
 
+    // @ts-ignore
```
There should be a way to type it so we don't need the ignore.

maxmellen replied:
Similarly to https://github.com/denoland/deno/pull/3899#discussion_r377604819, the time of the argument passed to `console[method]` still wouldn't be correct so I'm not sure how to improve on this.

maxmellen replied:
If I type `methods` as `Array<keyof typeof Console>` (apparently `Console` refers to a value in this context) I then get the following error:
```
error TS7053: Element implicitly has an 'any' type because expression of type 'string | number | symbol' can't be used to index type 'Console'.
  No index signature with a parameter of type 'string' was found on type 'Console'.

► file:///Users/max/Developer/deno/cli/js/console_test.ts:310:5

310     console[method]({
        ~~~~~~~~~~~~~~~
```
And even if we got over that error, then we'd have the issue that `{ toString(): void { hasCalled = true; } }` wouldn't be considered a valid argument to these functions anyway (for instance, `count` only accepts a string), so I think the best thing to do is switch `methods` back to implicitly be of type `string[]` as before my changes and keep the `@ts-ignore`.

kitsonk replied:
🤷‍♂ ok! you win! :-D

kitsonk commented on cli/js/console_test.ts:
```diff
@@ -464,7 +465,7 @@ test(function consoleGroupWarn(): void {
     console.warn("7");
     assertEquals(
-      both.toString(),
+      both!.toString(),
```
and `assert(both)` prior to this.

kitsonk commented on cli/js/console_test.ts:
```diff
@@ -694,6 +695,6 @@ test(function consoleDirXml(): void {
   mockConsole((console, _out, err): void => {
     console.trace("%s", "custom message");
-    assert(err.toString().includes("Trace: custom message"));
+    assert(err!.toString().includes("Trace: custom message"));
```
and `assert(err)` prior to this

kitsonk commented on cli/js/error_stack_test.ts:
```diff
@@ -1,8 +1,8 @@
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-const { setPrepareStackTrace } = Deno[Deno.symbols.internal] as any;
+// @ts-ignore
```
Was there a specific reason to replace this with `@ts-ignore`?

maxmellen replied:
Same issue as https://github.com/denoland/deno/pull/3899#discussion_r377584181

kitsonk commented on cli/js/event_target_test.ts:
```diff
@@ -4,8 +4,11 @@ import { test, assertEquals } from "./test_util.ts";
 test(function addEventListenerTest(): void {
   const document = new EventTarget();
 
+  // @ts-ignore
```
What are we ignoring here and ignoring.  Feels like something isn't right if this is failing.

maxmellen replied:
I have to admit that I was a bit confused by this test.
So with `strictNullChecks` turned on, `null` isn't considered a valid second argument to `document.addEventListener`.
I guess an alternative would be to cast `addEventListener` as `any` like this:
```typescript
  assertEquals((document.addEventListener as any)("x", null, false), undefined);
```
although I see that option as largely equivalent.

I'll gladly change the `// @ts-ignore` to `any` casts if that's what you'd prefer.

kitsonk commented on cli/js/event_target_test.ts:
```diff
@@ -34,11 +37,19 @@ test(function constructedEventTargetCanBeUsedAsExpected(): void {
-    on(type, callback?, options?): void {
+    on(
+      type: string,
+      callback: (e: Event) => void,
```
This was potion before the types.

maxmellen replied:
I'm sorry but I'm not sure what you mean

maxmellen replied:
Same issue as https://github.com/denoland/deno/pull/3899#discussion_r377594434

maxmellen replied:
Right so actually I guess it should be typed `(e: Event) => void | null` which is what TypeScript now expects with `strictNullChecks` and would also match TypeScript's DOM types.

maxmellen replied:
Right so actually I guess it should be typed `(e: Event) => void | null` which is what TypeScript now expects with `strictNullChecks` and would also match TypeScript's DOM types.

kitsonk commented on cli/js/event_target_test.ts:
```diff
@@ -60,8 +71,11 @@ test(function anEventTargetCanBeSubclassed(): void {
 
 test(function removingNullEventListenerShouldSucceed(): void {
   const document = new EventTarget();
+  // @ts-ignore
```
again, what are we ignoring and why?

maxmellen replied:
Same issue as https://github.com/denoland/deno/pull/3899#discussion_r377594434

kitsonk commented on cli/js/event_test.ts:
```diff
@@ -71,6 +71,7 @@ test(function eventPreventDefaultSuccess(): void {
 
 test(function eventInitializedWithNonStringType(): void {
   const type = undefined;
+  // @ts-ignore
```
I would force the type above to be any using an eslint ignore...

kitsonk commented on cli/js/event_test.ts:
```diff
@@ -85,11 +86,11 @@ test(function eventInitializedWithNonStringType(): void {
   const desc1 = Object.getOwnPropertyDescriptor(new Event("x"), "isTrusted");
   assertNotEquals(desc1, undefined);
-  assertEquals(typeof desc1.get, "function");
+  assertEquals(typeof desc1!.get, "function");
```
`assert(desc1)` prior to this

kitsonk commented on cli/js/fetch_test.ts:
```diff
@@ -96,6 +96,7 @@ testPerm({ net: true }, async function responseClone(): Promise<void> {
   const ab = await response.arrayBuffer();
   const ab1 = await response1.arrayBuffer();
   for (let i = 0; i < ab.byteLength; i++) {
+    // @ts-ignore
```
what are we ignoring and why?

maxmellen replied:
TypeScript doesn't consider indexing ArrayBuffers by number valid, I guess a better alternative might be to create Uint8Array views over these buffers and compare those.

kitsonk commented on cli/js/form_data_test.ts:
```diff
@@ -36,7 +36,9 @@ test(function formDataParamsGetSuccess(): void {
   formData.append("a", "true");
   formData.append("b", "false");
   formData.append("a", "null");
+  // @ts-ignore
```
IF these are values that don't throw, then they need to be included in the interfaces, not ignored.

maxmellen replied:
I would agree but as things currently are `FormData` in `cli/js/dom_types.ts` matches the definition in TypeScript's `lib/lib.dom.d.ts`. I personally wouldn't want to see Deno's DOM types diverge from TypeScript.
Again, I could alternatively change these to `any` casts,

What do you think?

kitsonk replied:
Yeah, we don't want to deviate, so commenting again of testing contra the typings.

kitsonk commented on cli/js/form_data_test.ts:
```diff
@@ -110,6 +114,7 @@ test(function formDataParamsArgumentsCheck(): void {
     let hasThrown = 0;
     let errMsg = "";
     try {
+      // @ts-ignore
```
There should be a way to type this do you don't have to ignore.

maxmellen replied:
The issue here is that `formData[method]` would take 1 or 2 arguments, which isn't what we do here, so I'm not sure how to circumvent that here.

kitsonk replied:
Understood.

kitsonk commented on cli/js/headers_test.ts:
```diff
@@ -22,14 +23,16 @@ test(function newHeaderTest(): void {
   name1: "value1",
   name2: "value2",
   name3: "value3",
+  // @ts-ignore
```
If undefined values are valid, the types should reflect this instead of ignoring it.

maxmellen replied:
These values ultimately end up being used to build a `Headers` object. Again, TypeScript's `lib/lib.dom.d.ts` definition for this type wouldn't accept `undefined` or `null`, so reflecting this would mean diverging from those types.

kitsonk replied:
Understood.

kitsonk commented on cli/js/os_test.ts:
```diff
@@ -249,7 +252,7 @@ testPerm({ env: true }, function getDir(): void {
       if (r.shouldHaveValue) {
         const d = Deno.dir(s.kind);
-        assert(d.length > 0);
+        assert(d!.length > 0);
```
`assert(d)` prior to this

kitsonk commented on cli/js/process_test.ts:
```diff
@@ -134,10 +134,10 @@ testPerm({ run: true }, async function runStdinPiped(): Promise<void> {
 
   const msg = new TextEncoder().encode("hello");
-  const n = await p.stdin.write(msg);
+  const n = await p.stdin!.write(msg);
```
`assert(p.stdin)` prior to this

kitsonk commented on cli/js/resources_test.ts:
```diff
@@ -39,8 +39,8 @@ testPerm({ read: true }, async function resourcesFile(): Promise<void> {
     Object.keys(resourcesBefore).length + 1
   );
-  const newRid = Object.keys(resourcesAfter).find((rid): boolean => {
+  const newRid = +Object.keys(resourcesAfter).find((rid): boolean => {
```
Why did this need to change functionally?

kitsonk replied:
Yeah, ok... I know `Object.keys()` doesn't always produce helpful types.

kitsonk commented on cli/js/timers_test.ts:
```diff
@@ -7,21 +7,22 @@ function deferred(): {
-  let resolve;
-  let reject;
-  const promise = new Promise((res, rej): void => {
+  let resolve: ((value?: {} | PromiseLike<{}>) => void) | undefined = undefined;
```
you should be able to just do `let resolve: ((value?: {} | PromiseLike<{}>) => void);`

kitsonk commented on cli/js/utime_test.ts:
```diff
@@ -19,8 +19,8 @@ testPerm({ read: true, write: true }, function utimeSyncFileSuccess(): void {
   const fileInfo = Deno.statSync(filename);
-  assertFuzzyTimestampEquals(fileInfo.accessed, atime);
-  assertFuzzyTimestampEquals(fileInfo.modified, mtime);
+  assertFuzzyTimestampEquals(fileInfo.accessed!, atime);
```
`assertFuzzyTimestampEquals` should accept a possibly undefined value on the left hand side...

kitsonk commented on std/fs/copy.ts:
```diff
@@ -22,18 +22,18 @@ async function ensureValidCopy(
   }
 
-  if (isCopyFolder && !destStat.isDirectory()) {
+  if (isCopyFolder && !destStat!.isDirectory()) {
```
`assert(destStat)` prior to this...

maxmellen replied:
Is it ok to use the `assert` from `std/testing/asserts.ts` here or should that function be copied into a `util` file like the one in `cli/js`?

kitsonk commented on std/fs/empty_dir_test.ts:
```diff
@@ -226,7 +226,7 @@ test(async function emptyDirPermission(): Promise<void> {
       });
 
-      const output = await Deno.readAll(stdout);
+      const output = await Deno.readAll(stdout!);
```
`assert(stdout)` prior to this

kitsonk commented on std/http/cookie.ts:
```diff
@@ -45,7 +45,7 @@ function toString(cookie: Cookie): string {
     out.push("HttpOnly");
   }
-  if (Number.isInteger(cookie.maxAge)) {
+  if (typeof cookie.maxAge == "number" && Number.isInteger(cookie.maxAge)) {
```
strict equals please

kitsonk commented on std/http/file_server.ts:
```diff
@@ -125,8 +125,8 @@ async function serveDir(
   for (const fileInfo of fileInfos) {
-    const filePath = posix.join(dirPath, fileInfo.name);
-    const fileUrl = posix.join(dirUrl, fileInfo.name);
+    const filePath = posix.join(dirPath, fileInfo.name || "");
```
```suggestion
    const filePath = posix.join(dirPath, fileInfo.name ?? "");
```

kitsonk commented on std/http/file_server.ts:
```diff
@@ -125,8 +125,8 @@ async function serveDir(
-    const filePath = posix.join(dirPath, fileInfo.name);
-    const fileUrl = posix.join(dirUrl, fileInfo.name);
+    const filePath = posix.join(dirPath, fileInfo.name || "");
+    const fileUrl = posix.join(dirUrl, fileInfo.name || "");
```
```suggestion
    const fileUrl = posix.join(dirUrl, fileInfo.name ?? "");
```

kitsonk commented on std/http/file_server.ts:
```diff
@@ -139,7 +139,7 @@ async function serveDir(
       mode: modeToString(fileInfo.isDirectory(), mode),
       size: fileInfo.isFile() ? fileLenToString(fileInfo.len) : "",
-      name: fileInfo.name,
+      name: fileInfo.name || "",
```
```suggestion
      name: fileInfo.name ?? "",
```

kitsonk commented on std/http/file_server.ts:
```diff
@@ -318,10 +318,10 @@ listenAndServe(
     } finally {
       if (CORSEnabled) {
-        setCORS(response);
+        setCORS(response!);
```
`assert(response)` prior to this

kitsonk commented on std/http/server_test.ts:
```diff
@@ -523,7 +523,7 @@ test(async function testReadRequestError(): Promise<void> {
       assertEquals(err, undefined);
       assertNotEquals(req, Deno.EOF);
-      for (const h of test.headers) {
+      for (const h of test.headers!) {
```
shouldn't be needed because of the assert???

kitsonk commented on std/node/global.ts:
```diff
@@ -1 +1,2 @@
+// @ts-ignore
 window["global"] = window;
```
```suggestion
globalThis["global"] = globalThis;
```

Not your problem, but I didn't realise this was there, also I suspect we need in here:

```ts
global {
  var global: typeof globalThis;
}
```

So that it can be augmented properly without needing to ignore.

kitsonk commented on std/node/module.ts:
```diff
@@ -384,7 +388,7 @@ class Module {
     Module._cache[filename] = module;
     if (parent !== undefined) {
-      relativeResolveCache[relResolveCacheIdentifier] = filename;
+      relativeResolveCache[relResolveCacheIdentifier!] = filename;
```
`assert(relResolveCacheIdentifier)` prior to this

kitsonk commented on std/path/win32.ts:
```diff
@@ -331,14 +331,14 @@ export function join(...paths: string[]): string {
   let slashCount = 0;
   assert(firstPart != null);
-  if (isPathSeparator(firstPart.charCodeAt(0))) {
+  if (isPathSeparator(firstPart!.charCodeAt(0))) {
```
the assertion should be working here so the not null assertion isn't needed...

kitsonk commented on std/util/deep_assign.ts:
```diff
@@ -24,7 +24,7 @@ export function deepAssign(
         target[key] = {};
       }
-      deepAssign(target[key] as Record<string, unknown>, value);
+      deepAssign(target[key] as Record<string, unknown>, value!);
```
`assert(value)` here or `deepAssign` should be accepting `undefined`.

kitsonk replied:
So we should be able to remove the not null assert?

maxmellen replied:
Yep, my bad :)

maxmellen commented:
@kitsonk I've updated my PR to take in account your comments. I've also raised questions/concerns to some of your comments.
I rebased my branch on master to stay up to date, and to keep things easy to review I've split the bulk of my work in 3 commits:
- 16ab35ce2cd19b3d3f5a9b354600adc1e54a68d9 is my work as you initially reviewed it rebased onto master with the git conflicts fixed
- f9eccc4ac5fdadb934f63f34d02cc5c8a8cd1cf1 addresses your review comments where I haven't raised questions/concerns of my own
- 3674746880a78d1df8fb7cdf97a7f73ff41b0453 adds support for strict mode in the code that was added to master since your review

bartlomieju commented:
@maxmellen awesome work; thank you! I resolved some minor conflict and gonna merge it once CI gets green.

maxmellen commented:
Hey @bartlomieju!
Just rebased and fixed the latest issues.
My changes since you had a look at my PR are contained in this commit: b499431542387e1f8ca5316cd3f97dfcc0291525.
CI passes but I'm not sure why the CLA assistant isn't reporting its status. You should be able to see that I signed it earlier.

bartlomieju commented:
@kitsonk can you take another look?

maxmellen commented:
Hey @bartlomieju, @kitsonk, I rebased my work on master and addressed your comments in my latest commit!
